File: LinkedList/LinkedListADT.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def create_node(self, data):
        return Node(data)

    def add_front(self, data):
        new_node = self.create_node(data)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node

    def add_back(self, data):
        new_node = self.create_node(data)
        if self.tail is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

    def delete_node(self, data):
        current = self.head
        previous = None

        while current is not None and current.data != data:
            previous = current
            current = current.next

        if current is None:
            print(f"Node with data {data} not found.")
            return

        if previous is None:
            self.head = current.next
        else:
            previous.next = current.next

        if current == self.tail:
            self.tail = previous

        del current

    def print_list(self):
        current = self.head
        while current is not None:
            print(f"{current.data} -> ", end="")
            current = current.next
        print("NULL")


if __name__ == "__main__":
    ll = LinkedList()

    ll.add_back(10)
    ll.add_back(20)
    ll.add_front(5)
    ll.add_front(1)
    ll.print_list()  # 1 -> 5 -> 10 -> 20 -> NULL

    ll.delete_node(10)
    ll.print_list()  # 1 -> 5 -> 20 -> NULL

    ll.delete_node(1)
    ll.print_list()  # 5 -> 20 -> NULL

File: LinkedList/LinkedListADT.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    int data;
    struct Node *next;
} node;

node *head = NULL;
node *tail = NULL;

node *createNode(int data){
    node *newNode = (node *)malloc(sizeof(node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

void add_front(int data){
    node *newNode = createNode(data);
    if(head==NULL){
        head = newNode;
        tail = newNode;
    }
    else{
        newNode->next = head;
        head = newNode;
    }
}

// 리스트 뒤에 노드를 추가하는 함수
void add_back(int data)
{
    node *new_node = createNode(data);
    if (tail == NULL)
    {
        head = new_node;
        tail = new_node;
    }
    else
    {
        tail->next = new_node;
        tail = new_node;
    }
}

// 특정 값을 가진 첫 번째 노드를 삭제하는 함수
void delete_node(int data)
{
    node *current = head;
    node *previous = NULL;

    while (current != NULL && current->data != data)
    {
        previous = current;
        current = current->next;
    }

    if (current == NULL)
    {
        printf("Node with data %d not found.\n", data);
        return;
    }

    if (previous == NULL)
    {
        head = current->next;
    }
    else
    {
        previous->next = current->next;
    }

    if (current == tail)
    {
        tail = previous;
    }

    free(current);
}

// 리스트를 출력하는 함수
void print_list()
{
    node *current = head;
    while (current != NULL)
    {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}
int main(){

    add_back(10);
    add_back(20);
    add_front(5);
    add_front(1);
    print_list(); // 1 -> 5 -> 10 -> 20 -> NULL

    delete_node(10);
    print_list(); // 1 -> 5 -> 20 -> NULL

    delete_node(1);
    print_list(); // 5 -> 20 -> NULL
}